README.md: add live examples

oleg-derevenetz opened the pull request:
Hi @platisd since I use your Action in a real project hosted on the GitHub, I want to offer to share some real-world examples of workflows used to run your Action, so that end users understand that running `clang-tidy` manually is not so difficult, and also so that they have a real example before their eyes instead of abstract `your-clang-tidy-script.sh` (and so they can even cut-and-paste it to use in their own workflows).

Also I would like to highlight the feature that allows to analyze PRs from forks in a safe manner, because it really is a "killer feature" of this Action. Other Actions are usually combined "all-in-one" and they just can't do it. And this is a very important feature for a popular GitHub project with many external contributors.

platisd commented on README.md:
```diff
@@ -57,25 +63,35 @@ jobs:
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y clang-tidy
+    - name: Prepare compile_commands.json
+      run: |
+        cmake -B build -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
```
Is it necessary to export the compile commands?
I've been [using it without](https://github.com/platisd/smartcar_shield/blob/master/.github/workflows/tests.yml).

oleg-derevenetz replied:
I don't use the `run-clang-tidy`, but at least `clang-tidy-diff` wants to have the `compile_commands.json` to know the build settings (defines, include paths and so on).

platisd replied:
OK I see, haven't used it myself. Can we document that `compile_commands.json` is necessary for `clang-tidy-diff` then and mention somewhere that `run-clang-tidy` is also an option? I am/was under the impression the latter is the go-to choice for running clang-tidy generally.

oleg-derevenetz replied:
I can offer the `run-clang-tidy` first before offering the `clang-tidy-diff`, but I don't have a working example, since I don't use it myself.

platisd replied:
Don't worry about a working example, just mention it in the text so the readers can google it.

platisd commented on README.md:
```diff
@@ -57,25 +63,35 @@ jobs:
+        mkdir clang-tidy-result
+    - name: Analyze
+      run: |
+        git diff -U0 HEAD^ | clang-tidy-diff -p1 -path build -export-fixes clang-tidy-result/fixes.yml
```
What does this do? Run `clang-tidy-diff` only for the files changed in the last commit or something? Needs more explanation if we are to add it to the README.

oleg-derevenetz replied:
> Run clang-tidy-diff only for the files changed in the last commit or something?

Yes, exactly. To not waste time running the `clang-tidy` on unchanged files. Very handy for pull requests.

platisd replied:
What happens when a pull request gets created containing multiple commits? Won't that limit the analysis only to the files modified in the last commit?

Also, since the Action is taking care of that, I'd say let's go for simplicity instead of speed/optimization.

oleg-derevenetz replied:
> What happens when a pull request gets created containing multiple commits? Won't that limit the analysis only to the files modified in the last commit?

This depends on the diff you pass to it. I use the following syntax:

```
git diff -U0 HEAD^ | clang-tidy-diff -p1 -path build -export-fixes clang-tidy-result/fixes.yml
```

```HEAD^``` means "first parent of the current branch", therefore, I compare the current version with the version from the parent commit of the entire branch, so I get all the changes at once, no matter how many commits there were.

> Also, since the Action is taking care of that, I'd say let's go for simplicity instead of speed/optimization.

I just don't use it, so I is not able to provide an exact example for the `run-clang-tidy` :)

platisd replied:
> I just don't use it, so I is not able to provide an exact example for the run-clang-tidy :)

I guess you are responding to a different comment, but just mention `run-clang-tidy` as an option (in the text before the sample yaml file) so the readers can look it up themselves if they need to.

oleg-derevenetz replied:
I feel that I have chosen the wrong wording regarding explanation of how the `clang-tidy-diff` works in case or PRs. Suppose we have this `master` branch:

```
A - B
```

and now we are creating a branch for a future pull request:

```
A - B [master]
    \
     C - D - E [pull request]
```

when the pull request is created, the following _potential_ merge commit `M` is created:

```
A - B - - - -  M[master]
    \         /
     C - D - E [pull request]
```

and `git diff -U0 HEAD^` means comparison between `M` and `B`, so we get all `C`, `D` and `E` commits in that diff.

platisd replied:
Yeap, got that, cool! :) I still think however that while it's a totally fine change, it adds a degree of complexity to an example that should be simple.

platisd replied:
If the `git diff -U0 HEAD^ |` is there only for optimization, can you please remove it from the examples? :angel:

platisd commented on README.md:
```diff
@@ -57,25 +63,35 @@ jobs:
+      run: |
+        git diff -U0 HEAD^ | clang-tidy-diff -p1 -path build -export-fixes clang-tidy-result/fixes.yml
+    - name: Run clang-tidy-pr-comments action
+      if: ${{ hashFiles( 'clang-tidy-result/fixes.yml' ) != '' }}
```
Is this making sure to run the Action only if there are fixes?
Doesn't the Action already handle that internally? Since this is an example, isn't it better to keep it simpler rather than optimized?

oleg-derevenetz replied:
It does - if this file exists, but is empty. But if it's not generated at all (and this can happen if the source files were changed in such a way that the code was only deleted, but not added or changed, like [here](https://github.com/ihhub/fheroes2/pull/3895)), Action just fails. So we need to run this Action only if the `fixes.yml` is present.

platisd replied:
Interesting edge-case! I am a little split over whether it should be excluded from the example to favor simplicity or modify the Action and let it handle missing files (i.e. not fail).

On the other hand, I am OK-ish including it in the example, however some more explanation is necessary (as to _why_ we need this check) either as a comment or in the `name` of the step.

What do you think?

oleg-derevenetz replied:
I can add a comment clarifying why this check is needed. I didn't modify the Action itself because there may be typos in the yml file name or path for example, and in this case it's better to catch it early (failing Action is preferred in this case).

platisd replied:
Sounds good!

platisd commented on README.md:
```diff
@@ -90,54 +106,151 @@ on: issue_comment
+        unzip clang-tidy-result.zip -d clang-tidy-result
+        echo "pr_id=$(cat clang-tidy-result/pr-id.txt)" >> $GITHUB_ENV
+        echo "pr_head_repo=$(cat clang-tidy-result/pr-head-repo.txt)" >> $GITHUB_ENV
+        echo "pr_head_ref=$(cat clang-tidy-result/pr-head-ref.txt)" >> $GITHUB_ENV
+    - uses: actions/checkout@v2
+      with:
+        repository: ${{ env.pr_head_repo }}
+        ref: ${{ env.pr_head_ref }}
+        persist-credentials: false
+    - name: Redownload analysis results
+      uses: actions/github-script@v3.1.0
+      with:
+        script: |
+          let artifacts = await github.actions.listWorkflowRunArtifacts({
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              run_id: ${{github.event.workflow_run.id }},
+          });
+          let matchArtifact = artifacts.data.artifacts.filter((artifact) => {
+              return artifact.name == "clang-tidy-result"
+          })[0];
+          let download = await github.actions.downloadArtifact({
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              artifact_id: matchArtifact.id,
+              archive_format: "zip",
+          });
+          let fs = require("fs");
+          fs.writeFileSync("${{github.workspace}}/clang-tidy-result.zip", Buffer.from(download.data));
+    - name: Extract analysis results
+      run: |
+        mkdir clang-tidy-result
+        unzip clang-tidy-result.zip -d clang-tidy-result
+    - name: Run clang-tidy-pr-comments action
+      if: ${{ hashFiles( 'clang-tidy-result/fixes.yml' ) != '' }}
+      uses: platisd/clang-tidy-pr-comments@master
+      with:
+        github_token: ${{ secrets.GITHUB_TOKEN }}
+        clang_tidy_fixes: clang-tidy-result/fixes.yml
+        pull_request_id: ${{ env.pr_id }}
```
This is getting very complex! :scream: Isn't it much easier to use [actions/download-artifact](https://github.com/actions/download-artifact)?

oleg-derevenetz replied:
> Isn't it much easier to use actions/download-artifact?

It is used here:

```
let download = await github.actions.downloadArtifact({
```

:) But it's not so trivial to download artifact **from another workflow**. `actions/download-artifact` is intended to be used to share artifacts **between different jobs in the same workflow**. If you want to grab an artifact from a different workflow, things are getting complicated.

platisd replied:
Aha I see! Very useful piece of code then regardless of this Action! :grin: 
I now have also realized what this workflow _actually_ does. :stuck_out_tongue: 

OK, I like it but README is getting very extensive. Can you add some Markdown (sub)headers for these cases and create a table of contents right before the `What` section for easier navigation?

oleg-derevenetz replied:
> OK, I like it but README is getting very extensive. Can you add some Markdown (sub)headers for these cases and create a table of contents right before the What section for easier navigation?

Yes, of course. I'll try to sort it out :)

platisd reviewed (commented):
Thanks for the contributions! I've left some comments, so let's discuss. :) 
Overall, we should keep the documentation focused on how to use the Action rather than how to use `clang-tidy`.

platisd commented on README.md:
```diff
@@ -19,18 +28,36 @@ leave a comment without blocking the pull request from being merged. It should f
-  itself and then, in both cases, specify the path where you would like the report to be exported.<br>
-  The very same path should be supplied to the GitHub Action.
+
+* It will **not** run `clang-tidy` for you. You are responsible for doing that and then supply the Action with
+  the path to your generated report (see [examples](#how) below). You can generate a
+  YAML report that includes *fixes* for a pull request using the following methods:
+
+  * Using the `run-clang-tidy` utility script with the `-export-fixes` argument. This script usually
+    comes with the `clang-tidy` packages. You can use it to run checks *for the entire codebase* of a
+    project at once.
+
+  * Using the `clang-tidy-diff` utility script with the `-export-fixes` argument. This script also usually
+    comes with the `clang-tidy` packages, and and it can be used to run checks only for code fragments that
+    *have been changed* in a specific pull request.
+
+  * Alternatively, you may use `--export-fixes` with `clang-tidy` itself in your own script.
```
:+1: :100:

oleg-derevenetz commented:
Hi @platisd I have addressed your comments, please take a look at your convenience.

> Overall, we should keep the documentation focused on how to use the Action rather than how to use `clang-tidy`.

On the one hand, this is true, but on the other... You see, when the end user looks at "he won't run clang-tidy for you" and abstract scripts like `your-clang-tidy-script.sh` in examples, he thinks: "OMG, it's all too complex, I have no idea how to run it all myself, I have no experience in writing bash, python and js scripts" (and he is right here - it's hard to make it all **properly** and **securely**), "so I'll just take it some all-in-one combined solution that will run clang-tidy for me" and then he will need to analyze 3rd-party pull requests and he runs the workflow under the pull_request_target permissions, and then... his repo gets pwned. Therefore, I think that it is better to provide working examples - maybe they look a bit complicated, but at least they work and they may be used as-is (with minimal changes) via cut-and-paste :)

oleg-derevenetz commented:
@platisd

> If the git diff -U0 HEAD^ | is there only for optimization, can you please remove it from the examples? 👼

You see, `clang-tidy-diff` expects diff as input, so I can't remove it :) I'll need to replace the `clang-tidy-diff` by something else "out of my head", and this will no longer be a working and tested example.

platisd commented:
> @platisd
> 
> > If the git diff -U0 HEAD^ | is there only for optimization, can you please remove it from the examples? angel
> 
> You see, `clang-tidy-diff` expects diff as input, so I can't remove it :) I'll need to replace the `clang-tidy-diff` by something else "out of my head", and this will no longer be a working and tested example.

Ooh, OK then, so it's not there just for optimization. :grin:
Will squash and merge :+1:
